contracts/FlowFees.cdc

```diff
@@ -33,6 +39,80 @@ pub contract FlowFees {
             emit TokensWithdrawn(amount: amount)
             return <-vault
         }
+
+        // Allows the administrator to change all the fee parameters at once
+        pub fun setFeeParameters(surgeFactor: UFix64, inclusionEffortCost: UFix64, executionEffortCost: UFix64) {
+            let newParameters = FeeParameters(surgeFactor: surgeFactor, inclusionEffortCost: inclusionEffortCost, executionEffortCost: executionEffortCost)
+            FlowFees.setFeeParameters(newParameters)
+        }
+
+        // Allows the administrator to change the fee surge factor
+        pub fun setFeeSurgeFactor(surgeFactor: UFix64) {
+            let oldParameters = FlowFees.getFeeParameters()
+            let newParameters = FeeParameters(surgeFactor: surgeFactor, inclusionEffortCost: oldParameters.inclusionEffortCost, executionEffortCost: oldParameters.executionEffortCost)
+            FlowFees.setFeeParameters(newParameters)
+        }
+    }
+
+    // A struct holding the fee parameters needed to calculate the fees
+    pub struct FeeParameters {
+        pub var surgeFactor: UFix64
+        pub var inclusionEffortCost: UFix64
+        pub var executionEffortCost: UFix64
+
+        init(surgeFactor: UFix64, inclusionEffortCost: UFix64, executionEffortCost: UFix64){
+            self.surgeFactor = surgeFactor
+            self.inclusionEffortCost = inclusionEffortCost
+            self.executionEffortCost = executionEffortCost
+        }
+    }
+
+    /// Called when a transaction is submitted to deduct the fee
+    /// from the AuthAccount that submitted it
+    pub fun deductTransactionFee(_ acct: AuthAccount, inclusionEffort: UFix64, executionEffort: UFix64) {
+        var feeAmount = self.computeFees(inclusionEffort: inclusionEffort, executionEffort: executionEffort)
+
+        if feeAmount == UFix64(0) {
+            // If there are no fees to deduct, do not continue, 
+            // so that there are no unnecessarily emitted events
+            return
+        }
+
+        let tokenVault = acct.borrow<&FlowToken.Vault>(from: /storage/flowTokenVault)
+            ?? panic("Unable to borrow reference to the default token vault")
+
+        
+        if feeAmount > tokenVault.balance {
+            // In the future this code path will never be reached, 
+            // as payers that are under account minimum balance will not have their transactions included in a collection
+            feeAmount = tokenVault.balance
```

Comment:
Won't this cause the transaction to fail because the vault will be below the minimum balance? Why not just fail with an error message here to be more clear?

Reply:
If the case is that the payers funds are to low to pay for the transaction, what we would ultimately want is for that transaction to fail much sooner (so it never even gets executed), but that is not in place yet.

The current logic is that if a payer sends a transaction that puts the payer below the minimum account balance, all the payers transactions after that will fail because the account is minimum account balance (unless the payer adds more FLOW to their vault). This means their account will have less and less FLOW until they hit 0.0.

This code is here because of the following case:

Lets say:
- payer P has 99 FLOW
- Minimum account balance is 100 FLOW
- and transaction fees are 5 FLOW per transaction

- after P sends 1 transaction that doesn't increase P's balance, P has 94 FLOW (the transaction failed)
- after P sends 19 transaction that doesn't increase P's balance, P has 4 FLOW (the transactions all failed)
- P sends 1 more transaction that doesn't increase P's balance. The choice is to say P's balance is less than the transaction fees, so don't deduct transaction fees (tx will still fail) **or** deduct the remaining 4 FLOW from P, so P is at 0

This code is the code for the second option.

Reply:
This is not in scope here, but that is still a problem that needs to be properly solved.

The current idea as I know it is that the payer needs to have enough funds on their account, to cover both inclusion fees (known without running the transaction) and the max execution fees (gas limit) set by the transaction. If there are not enough funds the transaction would never be included in a collection.

Reply:
thanks @janezpodhostnik, after reading FVM code, I just understood what you did here. 

I think there is still an edge case here (allowing 0 transaction fees)

- Alice has 100 FLOW on FlowTokenVault 
- Alice links <&FlowToken.Vault{FungibleToken.Balance}>(/public/flowTokenBalance) to this vault 
- Alice links <&FlowToken.Vault>(from: /storage/flowTokenVault) to an empty vault 
- Alice sends a transaction ( passes balance checks probably ) 
- Transaction fee deducted as 0 and succeeds
- Storage checks for updated Accounts succeeds according to : https://github.com/onflow/flow-core-contracts/blob/26afc84a19039ae7e6628b4bc4b4f53083be4abc/contracts/FlowStorageFees.cdc#L65

This is more in scope of FlowStorageFees I guess though.